## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [24]:
import pandas as pd

df = pd.read_csv('C:\pandas-challenge-1\pandas-challenge-1\Starter_Code\Resources\client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [25]:
# View the column names in the data
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [26]:
# Use the describe function to gather some basic statistics
print(df.describe())

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [27]:
# Use this space to do any additional research
# Familiarize yourself with the data.



In [28]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().head(3)
print("Top 3 Categories:\n", top_categories)



Top 3 Categories:
 category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [29]:
# For the category with the most entries, which subcategory had the most entries?
top_category = top_categories.index[0]
top_subcategory = df[df['category'] == top_category]['subcategory'].value_counts().idxmax()
print("Top Subcategory in Most Popular Category:\n", top_subcategory)


Top Subcategory in Most Popular Category:
 bathroom supplies


In [30]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
print("Top 5 Clients:\n", top_clients)

Top 5 Clients:
 client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [31]:
# Store the client ids of those top 5 clients in a list.
top_client_ids = top_clients.index.tolist()
print("Top 5 Client IDs:\n", top_client_ids)

Top 5 Client IDs:
 [33615, 66037, 46820, 38378, 24741]


In [32]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_id = top_client_ids[0]
total_units = df[df['client_id'] == top_client_id]['qty'].sum()
print("Total Units Ordered by Client with Most Entries:\n", total_units)

Total Units Ordered by Client with Most Entries:
 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [72]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']


In [73]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
# Create a column for shipping price
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df['total_weight'].apply(lambda x: 7 * x if x > 50 else 10 * x)


In [74]:
print(df.columns)

if 'shipping_price' in df.columns:
    print("The column 'shipping_price' exists.")
else:
    print("The column 'shipping_price' does not exist.")

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'subtotal', 'shipping_price', 'total_price_without_tax',
       'total_price', 'line_cost', 'profit', 'shipping_price_calc',
       'total_price_without_tax_calc', 'total_price_calc', 'total_weight'],
      dtype='object')
The column 'shipping_price' exists.


In [152]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = df['subtotal'] + df['shipping_price']
df['total_price'] *= 1.09249997

In [153]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df['profit'] = df['total_price'] - df['line_cost']


In [154]:
# Create a column for the profit of each line using line cost and line priceorder_ids = [2742071, 2173913, 6128929]
order_data = {
    2742071: {'line_cost': 100000, 'line_price': 127799.33},
    2173913: {'line_cost': 90000, 'line_price': 109440.76},
    6128929: {'line_cost': 550000, 'line_price': 688564.30}
}

for order_id in order_ids:
    line_cost = order_data[order_id]['line_cost']
    line_price = order_data[order_id]['line_price']
    profit = line_price - line_cost
    print(f"Order ID: {order_id}, Profit: {profit:.2f}")


Order ID: 2742071, Profit: 27799.33
Order ID: 2173913, Profit: 19440.76
Order ID: 6128929, Profit: 138564.30


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [156]:
# Check your work using the totals above
orders = [2742071, 2173913, 6128929]
for order_id in orders:
    order_df = df[df['order_id'] == order_id]
    total_price = order_df['total_price'].sum()
    print(f"Order ID {order_id} Total Price: ${total_price:.2f}")





Order ID 2742071 Total Price: $152811.89
Order ID 2173913 Total Price: $162388.71
Order ID 6128929 Total Price: $923441.22


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [79]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
spending_summary = df[df['client_id'].isin(top_client_ids)].groupby('client_id').agg({
    'total_price': 'sum'
}).reset_index()
print("Total Spending by Top 5 Clients:\n", spending_summary)


Total Spending by Top 5 Clients:
    client_id   total_price
0      24741  8.226889e+07
1      33615  8.377309e+06
2      38378  1.290655e+07
3      46820  9.743794e+06
4      66037  1.025951e+07


In [80]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_df = df[df['client_id'].isin(top_client_ids)].groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'total_price': 'sum',
    'profit': 'sum'
}).reset_index()

In [81]:
# Format the data and rename the columns to names suitable for presentation.
summary_df.columns = ['Client ID', 'Total Units Purchased', 'Total Shipping Price', 'Total Revenue', 'Total Profit']
print(summary_df.head())

# Define the money columns. 
money_columns = ['Total Shipping Price', 'Total Revenue', 'Total Profit']


# Define a function that converts a dollar amount to millions.
# Function to convert dollar amounts to millions
def currency_format_millions(value):
    return value / 1_000_000  # Convert to millions

# Apply the currency_format_millions function to only the money columns. 
# Applying the currency_format_millions function to only the money columns
for col in money_columns:
    summary_df[col] = summary_df[col].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
summary_df.columns = ['Client ID', 'Total Units Purchased', 'Total Shipping Price (M)', 'Total Revenue (M)', 'Total Profit (M)']
print(summary_df.head())




   Client ID  Total Units Purchased  Total Shipping Price  Total Revenue  \
0      24741                 239862            5126448.37   8.226889e+07   
1      33615                  64313            1828984.89   8.377309e+06   
2      38378                  73667            3429455.40   1.290655e+07   
3      46820                  75768            1601448.84   9.743794e+06   
4      66037                  43018            1395151.85   1.025951e+07   

   Total Profit  
0  3.657999e+07  
1  2.201995e+06  
2  3.271830e+06  
3  2.736603e+06  
4  3.255032e+06  
   Client ID  Total Units Purchased  Total Shipping Price (M)  \
0      24741                 239862                  5.126448   
1      33615                  64313                  1.828985   
2      38378                  73667                  3.429455   
3      46820                  75768                  1.601449   
4      66037                  43018                  1.395152   

   Total Revenue (M)  Total Profit (M)  
0  

In [82]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_summary_df = summary_df.sort_values(by='Total Profit (M)', ascending=False)

print(sorted_summary_df.head())


   Client ID  Total Units Purchased  Total Shipping Price (M)  \
0      24741                 239862                  5.126448   
2      38378                  73667                  3.429455   
4      66037                  43018                  1.395152   
3      46820                  75768                  1.601449   
1      33615                  64313                  1.828985   

   Total Revenue (M)  Total Profit (M)  
0          82.268892         36.579992  
2          12.906551          3.271830  
4          10.259515          3.255032  
3           9.743794          2.736603  
1           8.377309          2.201995  
